### PR TITLE
Implement on-demand skin downloader, remove Lappi's non-free skins (fixes #1166, fixes #2547)

### LINKED ISSUE #1166: Skin fetcher
From #1061 

Automatically get new skins, especially when other tees have an unrecognised skin by the local client, the local client can auto search or download etc.

### LINKED ISSUE #2547: CC-BY-NC-SA skin license
> lightbulb, tank, emo, bunny, hedgehog, napoleon, roman, caesar, nosey, savage:
> Copyright Lappi
> CC-BY-NC-SA license

I would argue that CC-BY-NC-SA is too restrictive. It certainly does not line up with the rest of ddnet's source and data licenses. Unless Lappi re-licenses, we should delete these skins.

### PR DESCRIPTION
**Pull request by def-**
Seems to work on first try, I'm somewhat surprised :D

Future ideas:

- I haven't checked how the performance impact is of a new skin joining,
  might want to move the loading to a separate thread?
- Use the skins from download directory instead of redownloading? Might
  want to compare modified time.
- Make all skins load on demand only to save some memory?

### CODE REVIEW

**Comment by Learath2** [comment]
Oh wow, this was planned for so long. Great to see it's finally getting there.

**Comment by Learath2** [comment]
- I guess we could look into threading this, but we should test first. It's probably not trivial to have 2 threads interacting with the graphics thread, even the graphics thread being separate is a bit of a headache.
- We can hash the skins and make a small request to check if the version in the skindb is different. If it is: we redownload.
- Loaing skins on-demand/lazily would be a great idea to implement now, as this patch is bound to increase the skin diversity.

**Comment by Jupeyy** [comment]
We dont need a 2nd thread to interact with the graphics backend:
Just create all required stuff for the skin and if finished, just send the preprepared texture to the GPU from the main thread(maybe limit the amount of skins loaded at once), since the graphic commands are executaed in order its safe to directly use the skin texture then.
same goes for on-demand skins then.

hashes sound good

and add an option to disable skin fetch

**Comment by Learath2** [comment]
@Jupeyy `g_Config.m_ClFetchSkins` is already there

File: src/game/client/components/skins.cpp [hunk 1]
```diff
@@ -226,5 +232,51 @@ int CSkins::FindImpl(const char *pName) const
...
+
+	CDownloadSkin Skin;
+	str_copy(Skin.m_aName, pName, sizeof(Skin.m_aName));
+	Skin.m_IsVanilla = false;
+	Skin.m_OrgTexture = m_aSkins[DefaultIndex].m_OrgTexture;
+	Skin.m_ColorTexture = m_aSkins[DefaultIndex].m_ColorTexture;
+	Skin.m_BloodColor = m_aSkins[DefaultIndex].m_BloodColor;
+
+	char aUrl[256];
+	str_format(aUrl, sizeof(aUrl), "https://ddnet.tw/skins/skin/%s.png", pName);
```
Comment (ardadem) [inline src/game/client/components/skins.cpp:276]: since we use downloadedmaps as path, downloadedskins would be more fit imo

Comment (def-) [reply src/game/client/components/skins.cpp:276]: Done

**Comment by Ekidoxx** [comment]
So this will automatically download all skins from DDNet's database on button click?

**Comment by def-** [comment]
Nope, downloads skins of other players as required, not all.

File: src/game/client/components/skins.h [hunk 2]
```diff
@@ -22,17 +23,25 @@ class CSkins : public CComponent
 		bool operator<(const CSkin &Other) { return str_comp(m_aName, Other.m_aName) < 0; }
 	};
 
+	struct CDownloadSkin : public CSkin
+	{
+		std::shared_ptr<CGetFile> pTask;
+		char aPath[MAX_PATH_LENGTH];
```
Comment (fokkonaut) [inline src/game/client/components/skins.h:29]: Member variables use `m_` prefix

File: src/game/client/components/skins.cpp (hunk 1 above)
Comment (fokkonaut) [reply src/game/client/components/skins.cpp:276]: This should be a config, people could have and provide their custom databases. As default this suits of course.

**Comment by fokkonaut** [comment]
We should have a config for the website, having `https://ddnet.tw/skins/skin` as default.

**Comment by def-** [comment]
> We should have a config for the website, having https://ddnet.tw/skins/skin as default because people can use their own databases then, or other ones than ddnets

Done.

> Loaing skins on-demand/lazily would be a great idea to implement now, as this patch is bound to increase the skin diversity.

The downloadedskins are always lazily loaded, never added to normal skins.

> We can hash the skins and make a small request to check if the version in the skindb is different. If it is: we redownload.

Data is not so big, so I don't want to implement this for this PR.

**Comment by fokkonaut** [comment]
I think the last `/` cutting off the skinname.png should not be included in the cfg instead should be in the code. `%s/%s.png`

**Comment by def-** [comment]
I disagree because some might want to implement dynamic skins, so you could do "https://foo.bar/skin?name=MyName". Imagine something like having your name written on the skin, or some randomized skin.

**Comment by def-** [comment]
@Learath2 got to try it yet since I fixed the nginx config?

**Comment by Learath2** [comment]
Works great, let's close #2547 with this too?

**Comment by def-** [comment]
Sure, you can remove the skins too.

**Comment by def-** [comment]
Well, or I can in this PR. Not much work.

**Comment by Learath2** [comment]
@def- conflict ^^

**Comment by def-** [comment]
fixed.

**Comment by Learath2** [comment]
LGTM, Thanks

bors r+

**Pull request merged**